caffeine@patapon.info/extension.js

```diff
@@ -127,12 +106,302 @@ const AppsTrigger = {
     ON_ACTIVE_WORKSPACE: 2
 };
 
-const CaffeineToggle = GObject.registerClass(
-class CaffeineToggle extends QuickSettings.QuickMenuToggle {
+const InhibitorManager = GObject.registerClass({
+    Signals: {
+        'update': {}
+    }
+}, class InhibitorManager extends GObject.Object {
+    _init(settings) {
+        super._init();
+
+        this._isInhibited = false;
+        this._inhibitorCookie = null;
+        this._userEnabled = false;
+        this._triggerApp = null;
+        this._tempManageLight = false;
+        this._lastReasons = [];
+        this._ignoredReasons = [];
+
+        // DBus proxies
+        this._sessionManager = new DBusSessionManagerProxy(Gio.DBus.session,
+            'org.gnome.SessionManager',
+            '/org/gnome/SessionManager');
+        this._colorProxy = new ColorProxy(
+            Gio.DBus.session,
+            'org.gnome.SettingsDaemon.Color',
+            '/org/gnome/SettingsDaemon/Color',
+            (proxy, error) => {
+                if (error) {
+                    log(error.message);
+                }
+            }
+        );
+
+        this._settings = settings;
+        this._appSystem = Shell.AppSystem.get_default();
+
+        // Update state when extension settings changed
+        this._settings.connectObject(
+            `changed::${SCREEN_BLANK}`,
+            () => this._forceUpdate(),
+            `changed::${FULLSCREEN_KEY}`,
+            () => this._updateState(),
+            `changed::${NIGHT_LIGHT_KEY}`,
+            () => this._updateState(),
+            `changed::${INHIBIT_APPS_KEY}`,
+            () => this._updateState(),
+            `changed::${TRIGGER_APPS_MODE}`,
+            () => this._updateState(), this);
+
+        // Update state when fullscreened
+        global.display.connectObject('in-fullscreen-changed', () => this._updateState(), this);
+
+        // Update when possible app triggers change
+        this._appSystem.connectObject('app-state-changed', () => this._updateState(), this);
+        global.display.connectObject('notify::focus-window', () => this._updateState(), this);
+        global.workspace_manager.connectObject('workspace-switched',
+            () => this._updateState(), this);
+
```

Comment:
Implemented relevant signals suggestion. I'm not convinced on the signal blocker, since if we have an event happen and then revert in a short space of time, the revert would be missed and it would be stuck in the old state. I think that's just going to cause bugs down the line.

Filtering the app id feels a little redundant, as it would be trying to do part of `_updateState()`'s job for it, I think it's better to leave it as a sort of sync method, that just takes in the environment around it when given a nudge.

I'm happy that this point is resolved, if you're not just let me know :)